packages/backend/src/core/preprocessing/Preprocessor.ts

```diff
@@ -0,0 +1,148 @@
+import { PreprocessedStateUpdateRepository } from '../../peripherals/database/PreprocessedStateUpdateRepository'
+import { StateUpdateRepository } from '../../peripherals/database/StateUpdateRepository'
+import { Logger } from '../../tools/Logger'
+
+export type SyncDirection = 'forward' | 'backward' | 'not-needed'
+
+export class Preprocessor {
+  constructor(
+    private preprocessedStateUpdateRepository: PreprocessedStateUpdateRepository,
+    private stateUpdateRepository: StateUpdateRepository,
+    private logger: Logger
+  ) {
+    this.logger = this.logger.for(this)
+  }
+
+  async sync() {
+    let direction: SyncDirection
+
+    do {
+      direction = await this.calculateRequiredSyncDirection()
+      if (direction === 'forward') {
+        await this.processNextStateUpdate()
+      } else if (direction === 'backward') {
+        await this.rollbackOneStateUpdate()
+      }
+    } while (direction !== 'not-needed')
+  }
+
+  // Preprocessor can move only one state-update forward or backward. This
+  // function returns the required direction of the next move.
+  //
+  // It's not enough to just compare the last processed-state-update id with the
+  // last state-update id because if reorg has happened, state-update id might
+  // have caught up and be the same (it's a sequential number).
+  //
+  // That's why state transition hash is also compared.  In general, this
+  // function moves backward until it finds a "common" state update, with the
+  // same id and state transition hash. Then it moves forward.
+  async calculateRequiredSyncDirection(): Promise<SyncDirection> {
+    const lastStateUpdate = await this.stateUpdateRepository.findLast()
+    const lastProcessedStateUpdate =
+      await this.preprocessedStateUpdateRepository.findLast()
+
+    // 1. Handle simple cases of empty state updates and empty preprocessed
+    // state updates.
+
+    if (lastStateUpdate === undefined) {
+      return lastProcessedStateUpdate === undefined ? 'not-needed' : 'backward'
+    }
+    if (lastProcessedStateUpdate === undefined) {
+      return 'forward'
+    }
+
+    // 2. Move back when current id is behind or there's hash mismatch,
+    // (due to reorg), or throw on missing state update.
+
+    if (lastStateUpdate.id < lastProcessedStateUpdate.stateUpdateId) {
+      return 'backward'
+    }
+
+    const correspondingStateUpdate = await this.stateUpdateRepository.findById(
```

Comment:
I have made a change to rely on `SyncStatusRepository.getLastSynced()` since it's the right place to get current sync state from, but your point stands that we don't have an explicit guarantee that some state update is not being processed while Preprocessor is working. Just like with collectors and SyncStatusRepository we will rely on a call order, i.e. we will call `Preprocessor.sync()` exactly after `SyncStatusRepository` is called to update last synced id (`.setLastSynced()`).

To have a proper guarantee we need to introduce a locking mechanism (`SELECT ... FOR UPDATE`). I don't think it's urgent right now so I added a Linear story to discuss it.